pyseir/rt/infer_rt.py

```diff
@@ -0,0 +1,617 @@
+
+    df = filter_and_smooth_input_data(
+        df=pd.DataFrame(dict(cases=observed_new_cases, deaths=observed_new_deaths), index=date),
+        include_deaths=include_deaths,
+    )
+    return df
+
+
+def filter_and_smooth_input_data(df: [pd.DataFrame], include_deaths=False) -> pd.DataFrame:
+    """Do Everything Strange Here Before it Gets to the Inference Engine"""
+    MIN_CUMULATIVE_COUNTS = dict(cases=20, deaths=10)
+    MIN_INCIDENT_COUNTS = dict(cases=5, deaths=5)
+
+    # Apply Business Logic To Filter Raw Data
+    for column in ["cases", "deaths"]:
+        requirements = [  # All Must Be True
+            df[column].count() > InferRtConstants.MIN_TIMESERIES_LENGTH,
+            df[column].sum() > MIN_CUMULATIVE_COUNTS[column],
+            df[column].max() > MIN_INCIDENT_COUNTS[column],
+        ]
+        # Now Apply Input Outlier Detection and Smoothing
+        filtered = utils.replace_outliers(df[column], log=rt_log)
+        assert len(filtered) == len(df[column])
+        smoothed = filtered.rolling(
+            InferRtConstants.COUNT_SMOOTHING_WINDOW_SIZE,
+            win_type="gaussian",
+            min_periods=InferRtConstants.COUNT_SMOOTHING_KERNEL_STD,
+            center=True,
+        ).mean(std=InferRtConstants.COUNT_SMOOTHING_KERNEL_STD)
+        # TODO: Only start once non-zero to maintain backwards compatibility?
+
+        # Check if the Post Smoothed Meets the Requirements
+        requirements.append(smoothed.max() > MIN_INCIDENT_COUNTS[column])
+
+        # Check include_deaths Flag
+        if column == "deaths" and not include_deaths:
+            requirements.append(False)
+        else:
+            requirements.append(True)
+
+        if all(requirements):
+            df[column] = smoothed
+            # TODO: save_input_data_smoothing() Figures
+        else:
+            df = df.drop(columns=column, inplace=False)
+            rt_log.info("Dropping:", columns=column, requirements=requirements)
+
+    return df
+
+
+class RtInferenceEngine:
+    """
+    This class extends the analysis of Bettencourt et al to include mortality data in a
+    pseudo-non-parametric inference of R_t.
+
+    Parameters
+    ----------
+    data: DataFrame
+        DataFrame with a Date index and at least one "cases" column.
+    include_deaths: bool
+        If True, include the deaths timeseries in the calculation. XCorrelated and Averaged
+    display_name: str
+        Needed for Figures. Should just return figures along with dataframe and then deal with title
+        and save location somewhere downstream.
+    fips: str
+        Just used for output paths. Should remove with display_name later.
+    """
+
+    def __init__(
+        self, data, display_name, fips, include_deaths=False,
+    ):
+
+        self.dates = data.index
+        self.cases = data.cases if "cases" in data else None
+        self.deaths = data.deaths if "deaths" in data else None
```

Comment:
Have you thought about what this class would look like if dates, cases, and deaths weren't instance variables? I'm just thinking about how best to separate out the class specific logic